modAionImpl/src/org/aion/zero/impl/sync/FastSyncManager.java

```diff
@@ -0,0 +1,138 @@
+package org.aion.zero.impl.sync;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.aion.base.util.ByteArrayWrapper;
+
+/**
+ * Directs behavior for fast sync functionality.
+ *
+ * @author Alexandra Roatis
+ */
+public final class FastSyncManager {
+
+    private boolean enabled;
+    private final AtomicBoolean complete = new AtomicBoolean(false);
+
+    private final Map<ByteArrayWrapper, byte[]> importedTrieNodes = new ConcurrentHashMap<>();
+
+    public FastSyncManager() {
+        this.enabled = false;
+    }
+
+    public void addImportedNode(ByteArrayWrapper key, byte[] value, DatabaseType dbType) {
+        if (enabled) {
+            importedTrieNodes.put(key, value);
+        }
+    }
+
+    public boolean containsExact(ByteArrayWrapper key, byte[] value) {
+        return enabled
+                && importedTrieNodes.containsKey(key)
+                && Arrays.equals(importedTrieNodes.get(key), value);
+    }
+
+    /** Changes the pivot in case of import failure. */
+    public void handleFailedImport(
+            ByteArrayWrapper key, byte[] value, DatabaseType dbType, int peerId, String peer) {
+        if (enabled) {
```

Comment:
Is there any valid reason to call this while enabled is false?

Reply:
The functionality is not fully defined there yet so I'm not entirely sure. I came across potential un-recoverable errors that will require disabling fast sync. That `if` is there more as a reminer that some methods may still be called while the message gets through to all threads. This class can be ignored for the moment, since the methods in it are now implemnted yet.